### PR TITLE
Add discount endpoints

### PR DESCRIPTION
**Pull request by rawform**
Added discount endpoints:
- POST api/discounts - CreateDiscount()
- GET api/discounts - GetAllDiscounts()
- GET api/discounts/{id} - GetDiscountById(id)
- PATCH api/discounts/{id} - UpdateDiscountValidUntil(id)
- DELETE api/discounts/{id} - ArchiveDiscount(id)
- POST api/discounts/{discountId}/item/{itemId} - UpdateDiscountOfItem(discountId, itemId)

Errors are all over the place right now (waiting for exception middleware)

Also:
- DateValidators
- Added BusinessId to Discount model

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -362,4 +361,3 @@ MigrationBackup/
 
 # Fody - auto-generated XML schema
 FodyWeavers.xsd
-OmgvaPOS/appsettings.json
```
Comment (AugustinasJarockis) [inline .gitignore:365]: Revert these changes.

File: OmgvaPOS/Program.cs [hunk 2]
```diff
@@ -15,6 +15,8 @@
 using OmgvaPOS.ItemManagement.Services;
 using OmgvaPOS.ItemVariationManagement.Repositories;
 using OmgvaPOS.ItemVariationManagement.Services;
+using OmgvaPOS.src.DiscountManagement.Service;
+using OmgvaPOS.src.DiscountManagement.Repository;
```
Comment (AugustinasJarockis) [inline OmgvaPOS/Program.cs:19]: Remove src from namespaces.

File: OmgvaPOS/appsettings.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  "ConnectionStrings": {
+    "LocalDatabase": "Data Source=localhost;Database=OmgvaPos;Integrated Security=false;User ID=omgva;Password=admin;TrustServerCertificate=True;"
+  },
+  "Jwt": {
+    "Issuer": "http://localhost:5260",
+    "Audience": "http://localhost:5260",
+    "SecretKey": "51fb27cdcec17756f4141dcc1c6cd45f009ee3ee7cc0a4de99c61a72fa88650a"
+  },
+  "UseCloudDatabase": false
+}
```
Comment (AugustinasJarockis) [inline OmgvaPOS/appsettings.json:18]: Delete this file.

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json.Linq;
+using OmgvaPOS.BusinessManagement.DTOs;
+using OmgvaPOS.DiscountManagement.Models;
+using OmgvaPOS.HelperUtils;
+using OmgvaPOS.src.DiscountManagement.DTOs;
+using OmgvaPOS.src.DiscountManagement.Mappers;
+using OmgvaPOS.src.DiscountManagement.Service;
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:9]: Remove src

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+        private readonly IDiscountService _discountService = discountService;
+        private readonly ILogger<DiscountController> _logger = logger;
+        //TODO: fix error handling once we get exception middleware
+        // hopefully with our own exceptions
+        // since I'd like to return correct HTTP response methods.
+        [HttpPost]
+        [Authorize(Roles = "Admin,Owner")]
+        [ProducesResponseType<DiscountDTO>(StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:34]: Missing 403 status code return

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public IActionResult CreateDiscount([FromBody] CreateDiscountRequest createDiscountRequest) {
+            long? businessId = JwtTokenHandler.GetTokenBusinessId(HttpContext.Request.Headers.Authorization!);
+            if (businessId == null)
+                return Forbid();
+            else
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:42]: Nitpick, but else is unneccesary

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public ActionResult<IEnumerable<DiscountDTO>> GetAllDiscounts() {
+            long? businessId = JwtTokenHandler.GetTokenBusinessId(HttpContext.Request.Headers.Authorization!);
+            if (businessId == null) return Forbid();
+
+            try {
+                List<DiscountDTO> discountDTOs;
+
+                if (HttpContext.User.IsInRole("Admin"))
+                    discountDTOs = _discountService.GetGlobalDiscounts();
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:76]: This endpooint is unnecessary

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            catch (Exception ex) {
+                _logger.LogError(ex, "An unexpected internal server error occured while getting all discounts.");
+                return StatusCode(StatusCodes.Status500InternalServerError, ex.Message);
+            }
+        }
+
+        [HttpGet("{id}")]
+        [Authorize(Roles = "Admin,Owner,Employee")]
+        [ProducesResponseType<DiscountDTO>(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:91]: Add 403 exception here.

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public IActionResult UpdateDiscountValidUntil([FromBody] DateTime newValidUntil, long id) {
+            if (!JwtTokenHandler.CanManageBusiness(HttpContext.Request.Headers.Authorization!, _discountService.GetDiscountNoException(id).BusinessId))
+                return Forbid();
+            // there is really nothing else you can update in a discount
+            // since we care about historical data:
+            // changing discount% or discount type would require creating a new discount all together
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:124]: I think we should do this, but probably in another PR

File: .gitignore (hunk 1 above)
Comment (rawform) [reply .gitignore]: Done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                _logger.LogError(ex, "An unexpected internal server error occured while updating the discount.");
+                return StatusCode(StatusCodes.Status500InternalServerError, ex.Message);
+            }
+        }
+
+        [HttpDelete("{id}")]
+        [Authorize(Roles = "Admin,Owner")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:139]: Needs 403 here also

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                _logger.LogError(ex, "An unexpected internal server error occured while updating the discount.");
+                return StatusCode(StatusCodes.Status500InternalServerError, ex.Message);
+            }
+        }
+
+        [HttpPost("{discountId}/item/{itemId}")]
+        [Authorize(Roles = "Admin,Owner,Employee")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:160]: 403 error

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        //TODO: does not work error in ItemRepository UpdateItem(Item item)
+        public IActionResult UpdateDiscountOfItem(long discountId, long itemId) {
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:163]: Gal pakeisti apply?

File: OmgvaPOS/src/DiscountManagement/DTOs/CreateDiscountRequest.cs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+﻿using OmgvaPOS.DiscountManagement.Enums;
+using System.ComponentModel.DataAnnotations;
+
+namespace OmgvaPOS.src.DiscountManagement.DTOs
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/DTOs/CreateDiscountRequest.cs:4]: Remove src

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        catch (Exception ex) {
+            _logger.LogError(ex, "An error occurred while saving discount.");
+            throw new ApplicationException("Error saving discount.");
+        }
+    }
+
+    public List<Discount> GetGlobalDiscounts() {
+        try {
+            return [.. _context.Discounts
+                .Where(d => d.Type == DiscountType.Item)];
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:34]: Add filter to filter out archived discounts

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        catch (Exception ex) {
+            _logger.LogError(ex, "An error occurred while fetching all discounts.");
+            throw new ApplicationException("Error fetching all discounts.");
+        }
+    }
+    public List<Discount> GetBusinessDiscounts(long businessId) {
+        try {
+            return [.. _context.Discounts
+                .Where(d => d.BusinessId == businessId)
+                .Where(d => d.Type == DiscountType.Item)];
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:45]: Filter archived discounts out here.

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            _logger.LogError(ex, $"An error occurred while retrieving item with ID {id}.");
+            throw new ApplicationException("Error retrieving item.");
+        }
+    }
+
+    public void UpdateDiscount(Discount discount) {
+        var existingDiscount = _context.Discounts.FirstOrDefault(d => d.Id == discount.Id);
+        // thats all we can update
+        existingDiscount.TimeValidUntil = discount.TimeValidUntil;
+        existingDiscount.IsArchived = discount.IsArchived;
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:67]: We should not give the ability to unarchive discounts, therefore this should not be allowed also.

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        _context.SaveChanges();
+    }
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:66]: There probably should be a separate method that archives discounts.Probably call it delete to stay consistent with others.

File: OmgvaPOS/src/DiscountManagement/DTOs/CreateDiscountRequest.cs (hunk 13 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/DTOs/CreateDiscountRequest.cs]: done

File: OmgvaPOS/Program.cs (hunk 2 above)
Comment (rawform) [reply OmgvaPOS/Program.cs]: done

File: OmgvaPOS/src/DiscountManagement/Service/DiscountService.cs [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        _discountRepository.AddDiscount(discount);
+        
+        //TODO:
+        if (request.Type == DiscountType.Order) {
+            throw new NotImplementedException("Order is not implemented yet");
+            //var order = _orderService.GetOrderById(request.OrderId);
+            //order.DiscountId = discount.Id;
+            //_orderService.UpdateOrder(order);
+        }
+
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Service/DiscountService.cs:51]: Discount may also be applied to order item. I am not sure if it is accounted for.

File: OmgvaPOS/appsettings.json (hunk 3 above)
Comment (rawform) [reply OmgvaPOS/appsettings.json]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 4 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs]: done

File: OmgvaPOS/src/DiscountManagement/Service/DiscountValidator.cs [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public static void ValidateDateCreate(DateTime dateTime) {
+        if (dateTime.Date < DateTime.Today) {
+            throw new ValidationException("A new discount must be valid at today's date.");
+        }
+    }
+
+    public static void ValidateDateUpdate(DateTime dateTime) {
+        if (dateTime.Date < DateTime.Today) {
+            throw new ValidationException("Cannot set \"Valid until:\" to a date in the past, to deactivate discount, archive it.");
```
Comment (AugustinasJarockis) [inline OmgvaPOS/src/DiscountManagement/Service/DiscountValidator.cs:17]: First and previous functions may be merged into one by passing the message.

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 5 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:26]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 8 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:83]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 10 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:131]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 11 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:152]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+        public IActionResult CreateDiscount([FromBody] CreateDiscountRequest createDiscountRequest) {
+            long? businessId = JwtTokenHandler.GetTokenBusinessId(HttpContext.Request.Headers.Authorization!);
+            if (businessId == null)
+                return Forbid();
+            else
+                createDiscountRequest.BusinessId = businessId;
+
+            try {
+                var discount = _discountService.CreateDiscount(createDiscountRequest);
+                DiscountDTO discountDTO = DiscountMapper.ToDTO(discount);
```
[the remark this answers is not in the document]
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs]: done

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 7 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:69]: removed

File: OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs (hunk 12 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Controller/DiscountController.cs:159]: Same endpoint is used for adding and removing discounts from an item.
This is done because in the frontend, discount is applied to items by clicking a checkbox on and off next to an item.

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs (hunk 14 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:31]: outdated

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs (hunk 15 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs]: done

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs (hunk 16 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs]: done

File: OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs (hunk 17 above)
Comment (rawform) [reply OmgvaPOS/src/DiscountManagement/Repository/DiscountRepository.cs:58]: done

**Review by AugustinasJarockis — APPROVED** [review]

**Review by MrMnt — APPROVED** [review]
